04.ls/ls.rb

```diff
@@ -2,26 +2,126 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 COLUMNS = 3
 TAB_WIDTH = 8
 
+ENTRY_TYPES = {
+  file: '-',
+  blockSpecial: 'b',
+  characterSpecial: 'c',
+  directory: 'd',
+  link: 'l',
+  fifo: 'p',
+  socket: 's',
+  unknown: 'w'
+}.freeze
+
+FILE_PERMISSIONS = {
+  '0': '---',
+  '1': '--x',
+  '2': '-w-',
+  '3': '-wx',
+  '4': 'r--',
+  '5': 'r-x',
+  '6': 'rw-',
+  '7': 'rwx'
+}.freeze
+
+HALF_A_YEAR_SECONDS = (60 * 60 * 24) * (365.2425 / 2)
+
 def main
-  file_names = search_file_names
+  options = ARGV.getopts('alr')
+  searched_file_names = options['a'] ? Dir['..', '.*', '*'] : Dir['*']
+  sorted_file_names = searched_file_names.sort_by(&:downcase)
+  file_names = options['r'] ? sorted_file_names.reverse : sorted_file_names
+  options['l'] ? display_long_format(file_names) : display_column_format(file_names)
+end
+
+def display_long_format(file_names)
+  long_formats = build_long_format_table(file_names)
+  total_blocks = long_formats.sum { |format| format[:blocks] }
+  puts "total #{total_blocks}"
+
+  widths = caluculate_max_length(long_formats)
+  long_formats.each do |format|
+    rows = [
+      "#{format[:file_mode]} ",
+      format[:hard_links].rjust(widths[:link_width]),
+      "#{format[:owner_name].ljust(widths[:owner_width])} ",
+      "#{format[:group_name].ljust(widths[:group_width])} ",
+      format[:byte_size].rjust(widths[:byte_width]),
+      format[:last_modified_time],
+      format[:path_name]
+    ]
+    puts rows.join(' ')
+  end
+end
+
+def build_long_format_table(file_names)
+  file_names.map do |file_name|
+    status = File.lstat(file_name)
+    {
+      file_mode: "#{ENTRY_TYPES[status.ftype.to_sym]}#{create_permissions(status)}",
+      hard_links: status.nlink.to_s,
+      owner_name: Etc.getpwuid(status.uid).name,
+      group_name: Etc.getgrgid(status.gid).name,
+      byte_size: status.blockdev? || status.chardev? ? format('%#x', status.rdev) : status.size.to_s,
+      last_modified_time: create_last_modified_time(status),
+      path_name: status.symlink? ? "#{file_name} -> #{File.readlink(file_name)}" : file_name,
+      blocks: status.blocks
+    }
+  end
+end
+
+def create_permissions(status)
+  octals = status.mode.to_s(8)[-3..].chars
+  special_bits_table = [
+    [status.setuid?, 's'],
+    [status.setgid?, 's'],
+    [status.sticky?, 't']
+  ]
+  octals.zip(special_bits_table).map do |octal, (is_special, character)|
+    standard_permissions = FILE_PERMISSIONS[octal.to_sym]
```

Comment:
```suggestion
    standard_permissions = FILE_PERMISSIONS[octal]
```

シンボルに変換するぐらいなら、FILE_PERMISSIONSのキーを最初から文字列にしておくのもありかと。

Reply:
to_sで文字列にしてからさらにto_symするのが冗長に感じたので、ハッシュのキーを文字列に変更しました。